fem/hyperbolic_conservation_laws.hpp

```diff
@@ -0,0 +1,779 @@
+                                   const IntegrationRule *ir = NULL)
+   {
+      mfem_error("NOT IMPLEMENTED");
+   }
+
+   /**
+    * @brief implement (F(u), grad v) with abstract F computed by ComputeFlux
+    *
+    * @param[in] el local finite element
+    * @param[in] Tr element transformation
+    * @param[in] elfun local coefficient of basis
+    * @param[out] elvect evaluated dual vector
+    */
+   virtual void AssembleElementVector(const FiniteElement &el,
+                                      ElementTransformation &Tr,
+                                      const Vector &elfun, Vector &elvect);
+
+   /**
+    * @brief implement <-hat(F)(u,x) n, [[v]]> with abstract hat(F) computed by
+    * ComputeFluxDotN and numerical flux object
+    *
+    * @param[in] el1 finite element of the first element
+    * @param[in] el2 finite element of the second element
+    * @param[in] Tr face element transformations
+    * @param[in] elfun local coefficient of basis from both elements
+    * @param[out] elvect evaluated dual vector <-hat(F)(u,x) n, [[v]]>
+    */
+   virtual void AssembleFaceVector(const FiniteElement &el1,
+                                   const FiniteElement &el2,
+                                   FaceElementTransformations &Tr,
+                                   const Vector &elfun, Vector &elvect);
+
+   virtual ~HyperbolicFormIntegrator() {}
+};
+
+
+// Base Hyperbolic conservation law class.
+// This contains all methods needed except the flux function.
+class DGHyperbolicConservationLaws : public TimeDependentOperator
+{
+private:
+   const int dim;
+   const int num_equations;
+   // Vector finite element space containing conserved variables
+   FiniteElementSpace *vfes;
+   // Element integration form. Should contain ComputeFlux
+   HyperbolicFormIntegrator *formIntegrator;
+   HyperbolicFormIntegrator *faceFormIntegrator;
+   // Base Nonlinear Form
+   NonlinearForm *nonlinearForm;
+   // element-wise inverse mass matrix
+   std::vector<DenseMatrix> Me_inv;
+   // global maximum characteristic speed. Updated by form integrators
+   mutable double max_char_speed;
+   // auxiliary variable used in Mult
+   mutable Vector z;
+
+   // Compute element-wise inverse mass matrix
+   void ComputeInvMass();
+
+   void Update()
+   {
+      nonlinearForm->Update();
+      height = nonlinearForm->Height();
+      width = height;
+      z.SetSize(height);
+
+      ComputeInvMass();
+   }
+
+public:
+   /**
+    * @brief Construct a new DGHyperbolicConservationLaws object
+    *
+    * @param vfes_ vector finite element space. Only tested for DG [Pₚ]ⁿ
+    * @param formIntegrator_ integrator (F(u,x), grad v)
+    * @param faceFormIntegrator_ integrator (F̂(u±, x, n), [[v]])
+    * @param num_equations_ the number of equations
+    */
+   DGHyperbolicConservationLaws(
+      FiniteElementSpace *vfes_,
+      HyperbolicFormIntegrator *formIntegrator_,
+      HyperbolicFormIntegrator *faceFormIntegrator_,
```

Comment:
Does it ever make sense to use different integrators for the elements and faces? I would think these should be the same

Reply:
`NonlinearForm` delete all integrators that are added to it. If we use an integrator for both domain and interior face integration, then the `~NonlinearForm` delete the integrator twice. I added a safeguard to avoid double deletion in the destructor `~DGHyperbolicConservationLaws`. I think this is not the best way to avoid having two `HyperbolicFormIntegrator`. But cannot think of any other simple solution.

Reply:
Good point. One option is to add `NonlinearForm::UseExternalIntegrators` just like `BilinearForm::UseExternalIntegrators`.